Feat: eas identifier attestation

micwallace opened the pull request:
Implements an EAS version of attestation.id identifier attestation, 
plus functionality to create and verify a ZKProof attestation from an EAS ticket and email attestation.

micwallace committed:
feat: EAS identifier attestation WIP

micwallace committed:
feat: EAS identifier attestation

- Fix tests
- Limit EasIdentifierAttestation construct to single key

micwallace committed:
feat: EAS identifier attestation

- Update packages to fix audit

micwallace committed:
Revert "feat: EAS identifier attestation"

This reverts commit bcb14ede03777dc9642f4e6d9ddd421ec54f66ba.

micwallace committed:
feat: EAS identifier attestation

- Fix audit

micwallace committed:
feat: A few extra functions and params

- Add public functions that return signer public key and address.
- Add param to skip revocation check when calling validateEasAttestation.

oleggrib commented on src/main/javascript/crypto/src/eas/EASIdentifierAttestation.ts:
```diff
@@ -0,0 +1,54 @@
+
+		if (!options.validity){
+			const date = new Date();
+			date.setDate(date.getDate() + 30);
```
better add input variable for TTL of attestation  in seconds so user developer can easy change it.
@micwallace

micwallace replied:
@oleggrib validity from and to date. This can be easily changed by the developer by passing these options. By default, for identifier attestations the validity is 30 days as per the current ASN implementation.

oleggrib commented on src/main/javascript/crypto/src/eas/EasAttestedObject.ts:
```diff
@@ -0,0 +1,112 @@
+import {UsageProofOfExponent} from "../libs/UsageProofOfExponent";
+import {DEBUGLEVEL} from "../config";
+
+export class EasUseToken {
```
would be great to move this ASN class to the file in /asn1 folder to they them in one place

micwallace replied:
I'd rather keep it here or as you said below, move to fully abi encoded structure.

oleggrib commented on src/main/javascript/crypto/src/eas/EasAttestedObject.ts:
```diff
@@ -0,0 +1,112 @@
+	getEncoded(){
+
+		const useToken = new EasUseToken();
+		useToken.ticketAttestation = new Uint8Array(this.ticketAttestation.getAsnEncoded());
```
@micwallace , no need to use ASN1 encoding in this case, better stay it solid and encode everything with defaultAbiCoder, like internal data encoded
in this case we dont need to use separate lib to decode ASN. it can save GAS in contract execution and save GAS when we deploy smaller contract.

micwallace replied:
Yep I agree. Keep in mind the contracts will require updating too. Can you help with that if I go ahead and make this change?

micwallace replied:
@oleggrib I've added an ABI encoded version. This is used by default but I've kept the ASN version around until the next major version.

oleggrib commented on src/main/javascript/crypto/src/eas/EASIdentifierAttestation.ts:
```diff
@@ -0,0 +1,54 @@
+		super(schema, signingConfig, undefined, issuerKey ? <KeysArray>{"": issuerKey} : undefined);
+	}
+
+	async createEasAttestation(data: EasIdentiferParameters|any, options?: EasTicketCreationOptions, commitmentType = 'mail'){
```
@micwallace commitment must be input value, it must be received from user side and commitmentSecret must stay in requestor side.

in current case some person, who has access to attestation.id privateKey can use it to issue attestation for ANY email. so its completely unsecure. Just check ASN version of idAttestation - it receive requestAttest and parse it for commitment and recovers userAddress

micwallace replied:
@oleggrib Yeah but surely if the user has attestation.id private key we have bigger problems! Is there any value in keeping this client-side. What if the server just generates a random secret and creates commitment for any email address? Generating secret client side doesn't seem to offer any additional protection. Server can always issue attestations to any address it wants.

oleggrib replied:
updated code looks great.

only the question in commitment+secret+userWalletAddress.

current attestation.id version generates commitment+secret in userBrowser and signs request. it solves few problems:
1. no chance to make mistake in walletAddress, because backend recovers walletAddress from signature. backend issues attestation to the wallet, what user really controls. no chance to use wrong/fake address by mistake

2. id attestation secret stores in user browser only and dont walk through the network. it means nobody can grab network data and get idAttestation

what do you think, @micwallace ?

micwallace replied:
Hey @oleggrib 

1. We still recover signature from Eip712AttestationRequest in order to set the correct eth address in the attestation, but this is done by common API instead of the library. You can review this code here: https://github.com/AlphaWallet/common-api/blob/85f9ec78a292c9f7cfc7547acd89f37c5509ff2a/backend/src/services/idAttestationsService.ts#L50C75-L50C75

2. We always use HTTPS so it's not possible to intercept the secret. In some ways it is better if there is no way for the client to specify the secret (and choose a weak one).

micwallace committed:
feat: add workflow dispatch and custom version option to snapshot creation

- Prevent "package not found" error when building token negotiator examples from snapshot

micwallace committed:
chore: change to StaticJsonRpcProvider to prevent unnecessary calls to eth_chainId

micwallace committed:
fix: Hex key import to support 0x prefix

micwallace committed:
feat: implement ABI encoding by default for UseTicketProof attestation bundle.

micwallace committed:
feat: implement ABI encoding by default for UseTicketProof attestation bundle.

micwallace committed:
chore: bump version

micwallace committed:
chore: bump version & fix schemaUID calc.

oleggrib reviewed (approved):
Thnak you, great, we can merge.
